cpp/ql/src/experimental/Security/CWE/CWE-570/WrongInDetectingAndHandlingMemoryAllocationErrors.ql

```diff
@@ -0,0 +1,93 @@
+/**
+ * @name Сonfusion In Detecting And Handling Memory Allocation Errors
+ * @description --::operator new(std::size_t) throws an exception on error, and ::operator new(std::size_t, const std::nothrow_t &) returns zero on error.
+ *              --the programmer can get confused when check the error that occurs when allocating memory incorrectly.
+ *              --Making a call of this type may result in a zero byte being written just outside the buffer.
+ * @kind problem
+ * @id cpp/detect-and-handle-memory-allocation-errors
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-570
+ */
+
+import cpp
+
+/**
+ * Lookup if condition compare with 0
+ */
+class IfCompareWithZero extends IfStmt {
+  IfCompareWithZero() {
+    this.getCondition().(EQExpr).getAChild().getValue() = "0"
+    or
+    this.getCondition().(NEExpr).getAChild().getValue() = "0" and
+    this.hasElse()
+  }
+}
+
+/**
+ * lookup for calls to `operator new`, with incorrect error handling.
+ */
+class WrongCheckErrorOperatorNew extends FunctionCall {
+  Expr exp;
+
+  WrongCheckErrorOperatorNew() {
+    this = exp.(NewOrNewArrayExpr).getAChild().(FunctionCall) and
+    (
+      this.getTarget().hasGlobalOrStdName("operator new")
+      or
+      this.getTarget().hasGlobalOrStdName("operator new[]")
+    )
+  }
+
+  /**
+   * Holds if handler `try ... catch` exists.
+   */
+  predicate isExistsTryCatchBlock() {
+    exists(TryStmt tb, AssignExpr aex, Initializer it |
+      tb.getAChild*() = exp
+      or
+      exp = it.getExpr() and
+      tb.getAChild*().(DeclStmt).getADeclaration() = it.getDeclaration()
+      or
+      aex.getAChild*() = exp and
+      tb.getAChild*().(AssignExpr) = aex
+    )
+  }
+
+  /**
+   * Holds if results call `operator new` check in `operator if`.
+   */
+  predicate isExistsIfCondition() {
+    exists(IfCompareWithZero ifc, AssignExpr aex, Initializer it |
+      // call `operator new` directly from the condition of `operator if`.
+      this = ifc.getCondition().getAChild()
+      or
+      // check results call `operator new` with variable appropriation
+      postDominates(ifc, this) and
+      aex.getAChild() = exp and
+      ifc.getCondition().getAChild().(VariableAccess).getTarget() =
+        aex.getLValue().(VariableAccess).getTarget()
+      or
+      // check results call `operator new` with declaration variable
+      postDominates(ifc, this) and
+      exp = it.getExpr() and
+      it.getDeclaration() = ifc.getCondition().getAChild().(VariableAccess).getTarget()
+    )
```

Comment:
This predicate contains quite a lot of code, I suspect we can come up with something simpler that may even cover more cases.  I'll have to come back to this though (it's getting late here).

Reply:
here's what I got with decreasing predicate `isExistsIfCondition`.
```ql
    exists(IfCompareWithZero ifc |
      globalValueNumber(exp) = globalValueNumber(ifc.getCondition().getAChild())
    )
```
the only negative is that in this case we will lose some detectors.
for example this kind.
https://lgtm.com/projects/g/axiomatic-systems/Bento4/snapshot/0c28aad63b6e5e5f78ed80cf62e71e430a26ae0f/files/Source/C++/Apps/Mp4Decrypt/Mp4Decrypt.cpp?sort=name&dir=ASC&mode=heatmap#L206
I suppose this is due to the initial configuration of the variable (but I could be wrong).

I would like to hear your opinion.

Reply:
Yep, the case in Bento4 is a shame but these cases are quite rare, and I definitely prefer it to the original solution.  I also tried this and it works quite well:
```
import semmle.code.cpp.dataflow.DataFlow

...

  exists(IfCompareWithZero ifc |
    DataFlow::localExprFlow(exp, ifc.getCondition().getAChild())
  )
```
It catches more than your `globalValueNumber` version, though there are a few strange cases due to flow around loops, e.g.
```
while (...)
{
	if (ptr == NULL)
	{
		...
	}

	ptr = new thing();
}
```

Reply:
Thank you for your offer.
unfortunately it adds false detection in loops and skips detection, for example such.  
https://lgtm.com/projects/g/wxWidgets/wxWidgets/snapshot/80b01b2d021e6354423a64cd33d17cac0a801542/files/src/common/imagiff.cpp?sort=name&dir=ASC&mode=heatmap#L342  
https://lgtm.com/projects/g/wxWidgets/wxWidgets/snapshot/80b01b2d021e6354423a64cd33d17cac0a801542/files/src/common/imagiff.cpp?sort=name&dir=ASC&mode=heatmap#L464  
https://lgtm.com/projects/g/alibaba/AliSQL/snapshot/69b5f9ebe36c28afffbe7a06ddd9ed629a5e0ab8/files/storage/innobase/row/row0import.cc?sort=name&dir=ASC&mode=heatmap#L1467  

The basic predicate works best of all, but taking into account the requirement for the elegance of the query, I suggest stopping at using `globalValueNumber`. In this case, we will lose part of the detection, but we will not add a false one.

Reply:
OK, if what we have is producing the best results I'm OK with merging it as-is.  It can always be improved in future.